### DISCUSSION TITLE
Subprocess not killed after clear all controls

### DISCUSSION THREAD

**Discussion by goodwin74** [cefsharp/CefSharp#3315] [Q&A]
When all items are removed from the Controls collection, the subprocess continues to run. Although the number of controls is zero.
Should it self-destruct or should I somehow kill the process manually after cleaning it?

**Comment by amaitland** [comment]
> Should it self-destruct

Only the render process associated with the `ChromiumWebBrowser` instance should be closed after you call `ChromiumWebBrowser.Dispose`.

The `GPU` process, network service (type utility) will pretty much always be active, even if you kill them they should automatically restart.

Which processes are running exactly? Check the `--type` param in Task Manager.

**Reply by goodwin74** [reply]
The problem is that in each tab, a CEF browser is created for me (and a separate process is started for it). After closing one tab and clearing all controls, the browser continues to work in the background (the process is visible in the Windows task manager). How to kill the process associated with this browser when closing a tab and clearing all controls?
browser.Dispose () - In my case, it completely closes my application :(

**Reply by amaitland** [reply]
> In my case, it completely closes my application

The default `CEF` behaviour is to send a `WM_Close` message when the browser is closed. Remove the `ChromiumWebBrowser` instance from it's parent before calling `Dispose` or you'll need to implement `ILifespanHandler`.

In future please include which version you are using and if you are using `WinForms/WPF/OffScreen` to make things easier for me to answer.

**Reply by goodwin74** [reply]
Sorry for not describing the platform. I am using WinForm CefSharp 86 version.

I just managed to close the process. Before closing the tab, I executed **"browser.GetBrowserHost (). CloseBrowser (true);"** and then cleared all controls. And everything worked, after closing the tab, the process associated with the browser ends.

**Reply by amaitland** [reply]
Disposing of the ChromiumWebBrowser instance will call CloseBrowser(true). You should only need to remove the ChromiumWebBrowser instance from it's parent when you call it's dispose method directly. 

Calling dispose on your tab control should dispose/remove all of the child controls. 
https://referencesource.microsoft.com/#System.Windows.Forms/winforms/Managed/System/WinForms/Control.cs,6034
